docs/designs/serving_design.md

```diff
@@ -0,0 +1,103 @@
+
+## Guarantee consistency between training and serving for data input
+
+ElasticDL is a distributed deep learning framework based on TensorFlow 2.0 eager execution. In ElasticDL, we use tf.data.Dataset to create input pipeline for training. The recommended way to preprocess data from tf.data.Dataset is to use [feature columns](https://www.tensorflow.org/tutorials/structured_data/feature_columns) in TensorFlow. What's more, tf.saved_model.save will save the defined feature columns with the model. So tf-serving will use the same preprocessing logic with training to make inference.
+
+Define a keras model with feature columns:
+
+```python
+def get_feature_columns():
+    age = tf.feature_column.numeric_column("age", dtype=tf.int64)
+    education = tf.feature_column.categorical_column_with_hash_bucket(
+        'education', hash_bucket_size=4)
+    education_one_hot = tf.feature_column.indicator_column(education)
+    return [age, education_one_hot]
+
+def get_input_layer():
+    input_layers = {}
+    input_layers['age'] = tf.keras.layers.Input(name='age', shape=(1,), dtype=tf.int64)
+    input_layers['education'] = tf.keras.layers.Input(name='education', shape=(1,), dtype=tf.string)
+    return input_layers
+
+def custom_model(feature_columns):
+    input_layers = get_input_layer()
+    dense_feature = tf.keras.layers.DenseFeatures(feature_columns=feat_cols)(input_layers)
+    dense = tf.keras.layers.Dense(10, activation='relu')(dense_feature)
+    dense = tf.keras.layers.Dense(1, activation='sigmoid')(dense)
+    return tf.keras.models.Model(inputs=input_layers, outputs=dense)
+```
+
+Although all feature columns in TensorFlow can be used in ElasticDL, tf.feature_column.embedding_column is not recommended in ElasticDL. Because the embedding_column has a variable containing a large embedding table. In eager execution the model must get all the embedding parameters to train. It will bring a large inter-process communication overhead.
+
+## Export the model with elasticdl.layers.Embedding to SavedModel
+
+Using native Keras layers to define a model is more user-friendly than using custom layers in ElasticDL. However, it is inefficient to train a model with tf.keras.layers.Embedding. When the model executes the forward-pass computation for each mini-batch, it must get all embedding parameters from the parameter server (PS) even if the mini-batch only contains several embedding ids. So, the elastic.layers.Embedding is designed to improve the training efficiency in ElasticDL. Considering user-friendliness and training efficiency, we need to define a model with tf.keras.layers.Embedding and train the model with elasticdl.layers.Embedding. For the Sequential model and the Functional models, we can use tf.keras.models.clone_model to replace the tf.keras.layers.Embedding with elasticdl.layers.Embedding before training starts. For subclass model, we can replace the attribute of tf.keras.layers.Embedding type with elastic.layers.Embedding.
```

Comment:
Keras does not have embedding layers with sparse inputs and a combiner. 
How to deal with such a layer?

Reply:
That's a good question. 
At the current stage, we use feature_column.embedding_column for sparse input.
But it can't leverage the benefit from the embedding sharding mechanism of  parameter server. We need figure out a way in a separate topic.